trulens_eval/trulens_eval/feedback/provider/endpoint/dummy.py

```diff
@@ -0,0 +1,388 @@
+from __future__ import annotations
+
+import inspect
+import logging
+from pprint import PrettyPrinter
+import random
+import sys
+from time import sleep
+from typing import Any, Callable, Dict, Optional, Sequence, Tuple, TypeVar
+
+from numpy import random as np_random
+import pydantic
+from pydantic import Field
+
+from trulens_eval.feedback.provider.endpoint.base import DEFAULT_RPM
+from trulens_eval.feedback.provider.endpoint.base import Endpoint
+from trulens_eval.feedback.provider.endpoint.base import EndpointCallback
+from trulens_eval.feedback.provider.endpoint.base import INSTRUMENT
+from trulens_eval.utils.python import locals_except
+from trulens_eval.utils.python import safe_hasattr
+from trulens_eval.utils.serial import JSON
+from trulens_eval.utils.threading import DEFAULT_NETWORK_TIMEOUT
+
+logger = logging.getLogger(__name__)
+
+pp = PrettyPrinter()
+
+A = TypeVar("A")
+B = TypeVar("B")
+T = TypeVar("T")
+
+
+class DummyAPI(pydantic.BaseModel):
+    """A dummy model evaluation API used by DummyEndpoint.
+
+    This is meant to stand in for classes such as OpenAI.completion . Methods in
+    this class are instrumented for cost tracking testing.
+    """
+
+    loading_prob: float
+    """How often to produce the "model loading" response that huggingface api
+    sometimes produces."""
+
+    loading_time: Callable[[DummyAPI],
+                           float] = lambda s: s.random.uniform(0.73, 3.7)
+    """How much time to indicate as needed to load the model."""
+
+    error_prob: float
+    """How often to produce an error response."""
+
+    freeze_prob: float
+    """How often to freeze instead of producing a response."""
+
+    overloaded_prob: float
+    """How often to produce the overloaded message that huggingface sometimes produces."""
+
+    alloc: int
+    """How much data in bytes to allocate when making requests."""
+
+    delay: float = 0.0
+    """How long to delay each request."""
+
+    seed: int = 0xdeadbeef
+    """Control randomness."""
+
+    random: Any = Field(exclude=True)
```

Comment:
Need to store the seeded random generator somewhere with users being this class and possibly other fake classes (like the custom_* classes in case they want to use randomness). There may be a better place to put it, perhaps in some "non-determinism controls for tests" class. Will think.

Have 2 as I used sampler functions from both python random and numpy random.

Reply:
Moved random samplers to a separate class to clean things up.